[Style] Use strong Style value types for size types width, height, min-width, min-height, max-width, max-height, and flex-basis

weinig opened the pull request:
#### 4bf6d54fbdaf53c88c9d5f7dc80acc19a9bf5b49
<pre>
[Style] Use strong Style value types for size types width, height, min-width, min-height, max-width, max-height, and flex-basis
<a href="https://bugs.webkit.org/show_bug.cgi?id=294041">https://bugs.webkit.org/show_bug.cgi?id=294041</a>

Reviewed by Darin Adler.

Adds types to represent &quot;preferred size&quot;, &quot;minimum size&quot;, &quot;maximum size&quot; and &quot;flex basis&quot;:

    Style::PreferredSize (wraps WebCore::Length) - used for the `width` and `height` properties.
    Style::MinimumSize (wraps WebCore::Length) - used for the `min-width` and `min-height` properties.
    Style::MaximumSize (wraps WebCore::Length) - used for the `max-width` and `max-height` properties.
    Style::FlexBasis (wraps WebCore::Length) - used for the `flex-basis` property.

The size types are purely zero-cost wrappers of `WebCore::Length`, meaning there
should be no runtime overhead here. The goal of the types is to enforce what
keywords (e.g. &apos;auto&apos;) and value ranges (all or nonnegative) are allowed, just
like with the recently added edge types.

Additionally, instead of allowing direct access to the underlying `WebCore::Length`
or to `WebCore::Length::value()`, the numeric value can be accessed via functions
like `tryFixed()` which return `std::optional&lt;Style::Length&lt;R&gt;&gt;` or via `WTF::switchOn`
which allows switch-statement like access to all alternatives. This means the
old idiomatic way for getting the value:

```
    if (size.isFixed())
        useValue(size.value());
```

is now replaced with:

```
    if (auto fixedValue = size.tryFixed())
        useValue(fixedValue-&gt;value);
```

or

```
    WTF::switchOn(size,
        [&amp;](const SizeType::Fixed&amp; fixedValue) {
            useValue(fixedValue.value);
        },
        [&amp;]...
    );
```

This ensures that the numeric value is not used without checking the type.

The biggest changes are in `RenderBox` and `RenderFlexibleBox` where functions that took
a `WebCore::Length` and a &quot;size type&quot; parameter now are overloaded for each relevant size
type. Where it made sense, a single template function implementation of the function was
used with overloads used only to allow the template function to stay out-of-line. In some
cases, however, the behavior for the different size types was distinct enough that just
using overloads was more appropriate. In some cases, overloads for each keyword term was
used. This allowed callers that know the exact type (say, `max-content`) to call directly
to the implementation for that type without going through the additional dispatch it used to.

* LayoutTests/editing/style/apply-style-atomic-expected.txt:
* LayoutTests/editing/style/apply-style-atomic-live-range-expected.txt:
    - Enforcement of nonnegative numeric range for width/height caught this
      issue where were setting the size to -100% for progress bars.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/style/values/flexbox/StyleFlexBasis.cpp: Added.
* Source/WebCore/style/values/flexbox/StyleFlexBasis.h: Added.
* Source/WebCore/style/values/sizing/StyleMaximumSize.cpp: Added.
* Source/WebCore/style/values/sizing/StyleMaximumSize.h: Added.
* Source/WebCore/style/values/sizing/StyleMinimumSize.cpp: Added.
* Source/WebCore/style/values/sizing/StyleMinimumSize.h: Added.
* Source/WebCore/style/values/sizing/StylePreferredSize.cpp: Added.
* Source/WebCore/style/values/sizing/StylePreferredSize.h: Added.
    - Add new files.

* Source/WebCore/style/values/box/StyleMargin.h:
* Source/WebCore/style/values/box/StylePadding.h:
* Source/WebCore/style/values/position/StyleInset.h:
* Source/WebCore/style/values/scroll-snap/StyleScrollMargin.h:
* Source/WebCore/style/values/scroll-snap/StyleScrollPadding.h:
    - Update existing Length wrapping style types to be consistent with
      new formatting and functions. A change in the future will unify these
      with a single base class so keeping them consistent is helpful.

* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
    - Update properties to use new types and associated builder/extractor functions.

* Source/WebCore/platform/Length.cpp:
* Source/WebCore/platform/Length.h:
    - Add support for using Length&apos;s empty value for WTF::HashTableEmptyValueType.
    - Rename isIntrinsicOrAuto() to isIntrinsicOrLegacyIntrinsicOrAuto() to make it
      more clear what it actual does.
    - Adds additional isIntrinsicOrLegacyIntrinsic() helper, for when auto is not
      wanted.
    - Added tryFixed() and tryPercentage() to allow using Length in generic contexts
      with the new style types.

* Source/WebCore/style/StyleResolver.cpp:
    - Ensure that we always have a root element style for the resolver state. More
      strict checking of length conversion caught this.

* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
    - Add missing template parameter needed for non-double Style::Percentage evaluation.

* Source/WebCore/style/values/primitives/StyleUnevaluatedCalculation.cpp:
* Source/WebCore/style/values/primitives/StyleUnevaluatedCalculation.h:
    - Add constructor for UnevaluatedCalculation that takes a CalculationValue&amp; to allow
      construction from Length without including CalculationValue.h in more places.

* Source/WebCore/css/values/primitives/CSSPrimitiveNumericUnits.h:
    - Adds support for multiplication of a unit literal with a scalar.
    - Adds support for division of a unit literal by a scalar.

* Source/WebCore/platform/Theme.cpp:
* Source/WebCore/platform/Theme.h:
* Source/WebCore/platform/adwaita/ThemeAdwaita.cpp:
* Source/WebCore/platform/adwaita/ThemeAdwaita.h:
* Source/WebCore/platform/mac/ThemeMac.h:
* Source/WebCore/platform/mac/ThemeMac.mm:
* Source/WebCore/rendering/RenderTheme.cpp:
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/adwaita/RenderThemeAdwaita.cpp:
* Source/WebCore/rendering/adwaita/RenderThemeAdwaita.h:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
* Source/WebCore/rendering/mac/RenderThemeMac.h:
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
    - Move controlSize() and minimumControlSize() to RenderTheme.

* Source/WebCore/dom/Document.cpp:
    - Removes variable `hasSpecifiedLogicalHeight` which was always false due to the expression
      `renderer-&gt;style().logicalMaxHeight().isAuto()` always being false. This was caught statically
      by the fact that `logicalMaxHeight()` now returns a `Style::MaximumSize` which has no `isAuto()`
      function on it.

* Source/WebCore/html/shadow/ProgressShadowElement.cpp:
    - Don&apos;t allow negative size values to be set here as the property is defined to take nonnegative lengths only.

* Source/WebCore/html/NumberInputType.cpp:
* Source/WebCore/html/SearchInputType.cpp:
* Source/WebCore/html/HTMLInputElement.cpp:
* Source/WebCore/html/HTMLTextFormControlElement.cpp:
* Source/WebCore/html/shadow/DateTimeNumericFieldElement.cpp:
* Source/WebCore/html/shadow/DateTimeSymbolicFieldElement.cpp:
* Source/WebCore/html/shadow/TextControlInnerElements.cpp:
* Source/WebCore/layout/formattingContexts/FormattingGeometry.cpp:
* Source/WebCore/layout/formattingContexts/FormattingGeometry.h:
* Source/WebCore/layout/formattingContexts/block/BlockFormattingGeometry.cpp:
* Source/WebCore/layout/formattingContexts/block/BlockFormattingQuirks.cpp:
* Source/WebCore/layout/formattingContexts/block/BlockMarginCollapse.cpp:
* Source/WebCore/layout/formattingContexts/flex/FlexFormattingContext.cpp:
* Source/WebCore/layout/formattingContexts/table/TableFormattingContext.cpp:
* Source/WebCore/layout/formattingContexts/table/TableFormattingQuirks.cpp:
* Source/WebCore/layout/integration/flex/LayoutIntegrationFlexLayout.cpp:
* Source/WebCore/layout/layouttree/LayoutElementBox.cpp:
* Source/WebCore/layout/layouttree/LayoutTreeBuilder.cpp:
* Source/WebCore/page/LocalFrameView.cpp:
* Source/WebCore/rendering/AutoTableLayout.cpp:
* Source/WebCore/rendering/AutoTableLayout.h:
* Source/WebCore/rendering/FixedTableLayout.cpp:
* Source/WebCore/rendering/FixedTableLayout.h:
* Source/WebCore/rendering/GridLayoutFunctions.cpp:
* Source/WebCore/rendering/GridMasonryLayout.cpp:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:
* Source/WebCore/rendering/PositionedLayoutConstraints.cpp:
* Source/WebCore/rendering/RenderAttachment.cpp:
* Source/WebCore/rendering/RenderBlock.cpp:
* Source/WebCore/rendering/RenderBlockFlow.cpp:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
* Source/WebCore/rendering/RenderBoxModelObject.h:
* Source/WebCore/rendering/RenderButton.cpp:
* Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/RenderFileUploadControl.cpp:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
* Source/WebCore/rendering/RenderFlexibleBox.h:
* Source/WebCore/rendering/RenderFragmentContainer.cpp:
* Source/WebCore/rendering/RenderGrid.cpp:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderListBox.cpp:
* Source/WebCore/rendering/RenderListMarker.cpp:
* Source/WebCore/rendering/RenderMenuList.cpp:
* Source/WebCore/rendering/RenderMultiColumnSet.cpp:
* Source/WebCore/rendering/RenderObject.cpp:
* Source/WebCore/rendering/RenderReplaced.cpp:
* Source/WebCore/rendering/RenderScrollbarPart.cpp:
* Source/WebCore/rendering/RenderSlider.cpp:
* Source/WebCore/rendering/RenderTable.cpp:
* Source/WebCore/rendering/RenderTable.h:
* Source/WebCore/rendering/RenderTableCell.cpp:
* Source/WebCore/rendering/RenderTableCell.h:
* Source/WebCore/rendering/RenderTableCellInlines.h:
* Source/WebCore/rendering/RenderTableSection.cpp:
* Source/WebCore/rendering/RenderTableSection.h:
* Source/WebCore/rendering/RenderTextControl.cpp:
* Source/WebCore/rendering/RenderTextControlMultiLine.cpp:
* Source/WebCore/rendering/RenderTextControlSingleLine.cpp:
* Source/WebCore/rendering/line/BreakingContext.h:
* Source/WebCore/rendering/mathml/RenderMathMLBlock.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleBoxData.cpp:
* Source/WebCore/rendering/style/StyleBoxData.h:
* Source/WebCore/rendering/style/StyleFlexibleBoxData.cpp:
* Source/WebCore/rendering/style/StyleFlexibleBoxData.h:
* Source/WebCore/rendering/style/TextSizeAdjustment.cpp:
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.cpp:
* Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
* Source/WebCore/style/StyleAdjuster.cpp:
* Source/WebCore/svg/SVGLengthContext.cpp:
* Source/WebCore/svg/SVGLengthContext.h:
    - Update for new types.

Canonical link: <a href="https://commits.webkit.org/296045@main">https://commits.webkit.org/296045@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5db29f3cb148dde1246dbf27fc66ecffaa11fc41

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107161 "105 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/26842 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17241 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112372 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/57712 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109125 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27513 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35346 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/81387 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/57712 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110090 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/21860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/96644 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61721 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/21284 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/14745 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/57175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/91215 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/14766 "Passed tests") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/115531 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34225 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/25254 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/115531 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34602 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/92869 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/115531 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22983 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35053 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/12831 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/29996 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34150 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/39683 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/33897 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37249 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35553 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/html/HTMLTextFormControlElement.cpp:
```diff
@@ -918,7 +918,7 @@ void HTMLTextFormControlElement::adjustInnerTextStyle(const RenderStyle& parentS
 
     if (parentStyle.fieldSizing() == FieldSizing::Content)
-        textBlockStyle.setLogicalMinWidth(Length { caretWidth(), LengthType::Fixed });
+        textBlockStyle.setLogicalMinWidth(Style::MinimumSize::Fixed { static_cast<float>(caretWidth()) });
```
I find casts like this vexing. Presumably we know this `int` fits in a `float` because a caret width can never be absurdly large, so maybe there should be a `floatCaretWidth` function where that comment can live?

weinig replied:
What do you think about just making my Fixed constructors a bit less strict and allowing them to take an int as well?

darinadler replied:
It continues the status quo so good for now; would have prevented me from noticing, but the underlying issues, perhaps inconsequential small ones, with mixing types likely remain the same.

darinadler commented on Source/WebCore/html/NumberInputType.cpp:
```diff
@@ -204,9 +204,13 @@ float NumberInputType::decorationWidth() const
         // So computedStyle()->logicalWidth() is used instead.
-        width += spinButton->computedStyle()->logicalWidth().value();
+
+        // FIXME: Handle non-fixed widths.
```
What does “handle” mean here? Could you make the comment more explicit about what’s wrong?

weinig replied:
Will do.

darinadler commented on Source/WebCore/html/SearchInputType.cpp:
```diff
@@ -195,10 +195,15 @@ bool SearchInputType::sizeShouldIncludeDecoration(int, int& preferredSize) const
-    if (RefPtr resultsButton = m_resultsButton; resultsButton && resultsButton->renderStyle())
-        width += m_resultsButton->renderStyle()->logicalWidth().value();
+    if (RefPtr resultsButton = m_resultsButton; resultsButton && resultsButton->renderStyle()) {
+        // FIXME: Handle non-fixed widths.
```
Should the three cases of this use a helper function so the fix is only in one place?

darinadler commented on Source/WebCore/html/SearchInputType.cpp:
```diff
@@ -195,10 +195,15 @@ bool SearchInputType::sizeShouldIncludeDecoration(int, int& preferredSize) const
+    }
     if (RefPtr cancelButton = m_cancelButton; cancelButton && cancelButton->renderStyle())
-        width += m_cancelButton->renderStyle()->logicalWidth().value();
+        // FIXME: Handle non-fixed widths.
```
Should the three cases of this use a helper function so the fix is only in one place?

darinadler commented on Source/WebCore/html/shadow/ProgressShadowElement.cpp:
```diff
@@ -92,7 +92,7 @@ ProgressValueElement::ProgressValueElement(Document& document)
 void ProgressValueElement::setInlineSizePercentage(double size)
 {
-    setInlineStyleProperty(CSSPropertyInlineSize, size, CSSUnitType::CSS_PERCENTAGE);
+    setInlineStyleProperty(CSSPropertyInlineSize, std::max(0.0, size), CSSUnitType::CSS_PERCENTAGE);
```
Should there be a 100.0 maximum too? If not, why not?

weinig replied:
The type is [0, inf]. It has no problem with values greater than 100%. Just values less than 0.

darinadler commented on Source/WebCore/html/shadow/TextControlInnerElements.cpp:
```diff
@@ -149,7 +146,7 @@ std::optional<Style::UnadjustedStyle> TextControlInnerElement::resolveCustomStyl
         auto emSize = CSSPrimitiveValue::create(1, CSSUnitType::CSS_EM);
         int pixels = emSize->resolveAsLength<int>(CSSToLengthConversionData { *newStyle, nullptr, nullptr, nullptr });
-        newStyle->setFlexBasis(Length { pixels, LengthType::Fixed });
+        newStyle->setFlexBasis(Style::FlexBasis::Fixed { static_cast<float>(pixels) });
```
Maybe we need a named function for this conversion to `float` that expresses why it’s safe? Or maybe we should not be doing a `resolveAsLength<int>` just to immediately cast to `float` and then convert to `Fixed`? I just hate to have a bare `static_cast<float>` without a rationale.

darinadler commented on Source/WebCore/layout/formattingContexts/block/BlockMarginCollapse.cpp:
```diff
@@ -275,7 +275,7 @@ bool BlockMarginCollapse::marginAfterCollapsesWithParentMarginAfter(const Elemen
         return false;
 
-    // Only the last inlflow child collapses with parent.
+    // Only the last inflow child collapses with parent.
```
Is inflow child exactly what the CSS specification calls this? If practical would love to fix more than the typo, and make it use just the right language.

darinadler commented on Source/WebCore/layout/formattingContexts/block/BlockMarginCollapse.cpp:
```diff
@@ -385,8 +384,7 @@ bool BlockMarginCollapse::marginsCollapseThrough(const ElementBox& layoutBox) co
     auto& style = layoutBox.style();
-    auto computedHeightValueIsZero = style.height().isFixed() && !style.height().value();
-    if (!(style.height().isAuto() || computedHeightValueIsZero))
+    if (auto& height = style.height(); !(height.isAuto() || (height.isFixed() && height.isZero())))
```
Can `height.isFixed() && height.isZero()` ever return something different from just `height.isZero()`? Maybe `isZero` asserts if called on a non-fixed value?

weinig replied:
If height is a percentage it would be different. Now, should 0% behave different than 0px here, I don’t know. (Usually the answer is no though).

darinadler commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -4446,7 +4446,7 @@ void LocalFrameView::performFixedWidthAutoSize()
     ASSERT(is<RenderElement>(*firstChild));
     auto& documentRenderer = downcast<RenderElement>(*firstChild);
-    documentRenderer.mutableStyle().setMaxWidth(Length(m_autoSizeConstraint.width(), LengthType::Fixed));
+    documentRenderer.mutableStyle().setMaxWidth(Style::MaximumSize::Fixed { static_cast<float>(m_autoSizeConstraint.width()) });
```
I have the same feelings here about the cast from `int` to `float` that I mentioned above.

darinadler commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -2335,12 +2335,12 @@ void RenderBlock::computePreferredLogicalWidths()
-        m_minPreferredLogicalWidth = adjustContentBoxLogicalWidthForBoxSizing(lengthToUse);
+    auto& styleToUse = style();
+    auto logicalWidth = overridingLogicalWidthForFlexBasisComputation().value_or(styleToUse.logicalWidth());
+    if (auto fixedLogicalWidth = logicalWidth.tryFixed(); !isRenderTableCell() && fixedLogicalWidth && fixedLogicalWidth->value >= 0 && !(isDeprecatedFlexItem() && !static_cast<int>(fixedLogicalWidth->value))) {
```
What makes the `static_cast<int>` safe? If nothing, maybe make that check into a helper function so we can try to have a comment where it explains itself?

darinadler commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -4610,13 +4609,16 @@ static inline LayoutUnit preferredWidth(LayoutUnit preferredWidth, float result)
 {
     auto& style = renderer.style();
+    if (auto fixedTextIndent = style.textIndent().tryFixed())
+        return fixedTextIndent->value ? std::make_optional(LayoutUnit { fixedTextIndent->value }) : std::nullopt;
```
Is this conversion from one `std::optional` to another something we should have a helper function for?

darinadler commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -547,7 +547,13 @@ static void updateSliderTrackPartForRenderer(SliderTrackPart& sliderTrackPart, c
+
+        auto fixedWidth = thumbStyle.width().tryFixed();
+        auto fixedHeight = thumbStyle.height().tryFixed();
+        auto thumbWidth = fixedWidth ? static_cast<int>(fixedWidth->value) : 0;
+        auto thumbHeight = fixedHeight ? static_cast<int>(fixedHeight->value) : 0;
```
Got the same unpleasant feeling here when we do `static_cast<int>`, like why are we truncating? Why is this right? Not that I want you to change it, just that I don’t love the idiom.

darinadler commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1417,18 +1428,62 @@ void RenderTheme::adjustButtonOrCheckboxOrColorWellOrInnerSpinButtonOrRadioStyle
+            if (fixedOverrideMinWidth->value > fixedOriginalMinWidth->value)
+                style.setMinWidth(Style::MinimumSize(minimumControlSize.width()));
+        } else if (auto percentageOriginalMinWidth = style.minWidth().tryPercentage()) {
+            // FIXME: This really makes no sense but matches existing behavior. Should use a `calc(max(override, original))` here instead.
```
To state the obvious: Would be so cool to have test cases for this.

darinadler commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1417,18 +1428,62 @@ void RenderTheme::adjustButtonOrCheckboxOrColorWellOrInnerSpinButtonOrRadioStyle
+        }
+    } else if (auto percentageOverrideMinWidth = minimumControlSize.width().tryPercentage()) {
+        if (auto fixedOriginalMinWidth = style.minWidth().tryFixed()) {
+            // FIXME: This really makes no sense but matches existing behavior. Should use a `calc(max(override, original))` here instead.
```
To state the obvious: Would be so cool to have test cases for this.

darinadler commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1417,18 +1428,62 @@ void RenderTheme::adjustButtonOrCheckboxOrColorWellOrInnerSpinButtonOrRadioStyle
+            if (fixedOverrideMinHeight->value > fixedOriginalMinHeight->value)
+                style.setMinHeight(Style::MinimumSize(minimumControlSize.height()));
+        } else if (auto percentageOriginalMinHeight = style.minHeight().tryPercentage()) {
+            // FIXME: This really makes no sense but matches existing behavior. Should use a `calc(max(override, original))` here instead.
```
To state the obvious: Would be so cool to have test cases for this.

darinadler commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1417,18 +1428,62 @@ void RenderTheme::adjustButtonOrCheckboxOrColorWellOrInnerSpinButtonOrRadioStyle
+        }
+    } else if (auto percentageOverrideMinHeight = minimumControlSize.height().tryPercentage()) {
+        if (auto fixedOriginalMinHeight = style.minHeight().tryFixed()) {
+            // FIXME: This really makes no sense but matches existing behavior. Should use a `calc(max(override, original))` here instead.
```
To state the obvious: Would be so cool to have test cases for this.

darinadler commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1417,18 +1428,62 @@ void RenderTheme::adjustButtonOrCheckboxOrColorWellOrInnerSpinButtonOrRadioStyle
+
+    // FIXME: The min-width/min-heigh value should use `calc-size()` when supported to make non-specified overrides work.
+
+    if (auto fixedOverrideMinWidth = minimumControlSize.width().tryFixed()) {
```
Is there a way to make the code below less repetitive? Helper functions perhaps?

darinadler commented on Source/WebCore/svg/SVGLengthContext.cpp:
```diff
@@ -102,19 +104,18 @@ static inline float dimensionForLengthMode(SVGLengthMode mode, FloatSize viewpor
 float SVGLengthContext::valueForLength(const Length& length, SVGLengthMode lengthMode)
```
When can we delete this one? Never? Soon?

weinig replied:
I hope soon. Got to get through a few more.

darinadler commented on Source/WebCore/svg/SVGLengthContext.cpp:
```diff
@@ -123,6 +124,29 @@ float SVGLengthContext::valueForLength(const Length& length, SVGLengthMode lengt
+        [&](const auto&) -> float {
+            return 0;
+
+        }
```
```suggestion
        [&](const auto&) -> float {
            return 0;
        }
```

darinadler reviewed (approved):
This change draws my attention to our existing `int`/`float`/`LayoutUnit` conversions, and how chaotic they seem.

weinig committed:
[Style] Use strong Style value types for size types width, height, min-width, min-height, max-width, max-height, and flex-basis
https://bugs.webkit.org/show_bug.cgi?id=294041

Reviewed by Darin Adler.

Adds types to represent "preferred size", "minimum size", "maximum size" and "flex basis":

    Style::PreferredSize (wraps WebCore::Length) - used for the `width` and `height` properties.
    Style::MinimumSize (wraps WebCore::Length) - used for the `min-width` and `min-height` properties.
    Style::MaximumSize (wraps WebCore::Length) - used for the `max-width` and `max-height` properties.
    Style::FlexBasis (wraps WebCore::Length) - used for the `flex-basis` property.

The size types are purely zero-cost wrappers of `WebCore::Length`, meaning there
should be no runtime overhead here. The goal of the types is to enforce what
keywords (e.g. 'auto') and value ranges (all or nonnegative) are allowed, just
like with the recently added edge types.

Additionally, instead of allowing direct access to the underlying `WebCore::Length`
or to `WebCore::Length::value()`, the numeric value can be accessed via functions
like `tryFixed()` which return `std::optional<Style::Length<R>>` or via `WTF::switchOn`
which allows switch-statement like access to all alternatives. This means the
old idiomatic way for getting the value:

```
    if (size.isFixed())
        useValue(size.value());
```

is now replaced with:

```
    if (auto fixedValue = size.tryFixed())
        useValue(fixedValue->value);
```

or

```
    WTF::switchOn(size,
        [&](const SizeType::Fixed& fixedValue) {
            useValue(fixedValue.value);
        },
        [&]...
    );
```

This ensures that the numeric value is not used without checking the type.

The biggest changes are in `RenderBox` and `RenderFlexibleBox` where functions that took
a `WebCore::Length` and a "size type" parameter now are overloaded for each relevant size
type. Where it made sense, a single template function implementation of the function was
used with overloads used only to allow the template function to stay out-of-line. In some
cases, however, the behavior for the different size types was distinct enough that just
using overloads was more appropriate. In some cases, overloads for each keyword term was
used. This allowed callers that know the exact type (say, `max-content`) to call directly
to the implementation for that type without going through the additional dispatch it used to.

* LayoutTests/editing/style/apply-style-atomic-expected.txt:
* LayoutTests/editing/style/apply-style-atomic-live-range-expected.txt:
    - Enforcement of nonnegative numeric range for width/height caught this
      issue where were setting the size to -100% for progress bars.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/style/values/flexbox/StyleFlexBasis.cpp: Added.
* Source/WebCore/style/values/flexbox/StyleFlexBasis.h: Added.
* Source/WebCore/style/values/sizing/StyleMaximumSize.cpp: Added.
* Source/WebCore/style/values/sizing/StyleMaximumSize.h: Added.
* Source/WebCore/style/values/sizing/StyleMinimumSize.cpp: Added.
* Source/WebCore/style/values/sizing/StyleMinimumSize.h: Added.
* Source/WebCore/style/values/sizing/StylePreferredSize.cpp: Added.
* Source/WebCore/style/values/sizing/StylePreferredSize.h: Added.
    - Add new files.

* Source/WebCore/style/values/box/StyleMargin.h:
* Source/WebCore/style/values/box/StylePadding.h:
* Source/WebCore/style/values/position/StyleInset.h:
* Source/WebCore/style/values/scroll-snap/StyleScrollMargin.h:
* Source/WebCore/style/values/scroll-snap/StyleScrollPadding.h:
    - Update existing Length wrapping style types to be consistent with
      new formatting and functions. A change in the future will unify these
      with a single base class so keeping them consistent is helpful.

* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
    - Update properties to use new types and associated builder/extractor functions.

* Source/WebCore/platform/Length.cpp:
* Source/WebCore/platform/Length.h:
    - Add support for using Length's empty value for WTF::HashTableEmptyValueType.
    - Rename isIntrinsicOrAuto() to isIntrinsicOrLegacyIntrinsicOrAuto() to make it
      more clear what it actual does.
    - Adds additional isIntrinsicOrLegacyIntrinsic() helper, for when auto is not
      wanted.
    - Added tryFixed() and tryPercentage() to allow using Length in generic contexts
      with the new style types.

* Source/WebCore/style/StyleResolver.cpp:
    - Ensure that we always have a root element style for the resolver state. More
      strict checking of length conversion caught this.

* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
    - Add missing template parameter needed for non-double Style::Percentage evaluation.

* Source/WebCore/style/values/primitives/StyleUnevaluatedCalculation.cpp:
* Source/WebCore/style/values/primitives/StyleUnevaluatedCalculation.h:
    - Add constructor for UnevaluatedCalculation that takes a CalculationValue& to allow
      construction from Length without including CalculationValue.h in more places.

* Source/WebCore/css/values/primitives/CSSPrimitiveNumericUnits.h:
    - Adds support for multiplication of a unit literal with a scalar.
    - Adds support for division of a unit literal by a scalar.

* Source/WebCore/platform/Theme.cpp:
* Source/WebCore/platform/Theme.h:
* Source/WebCore/platform/adwaita/ThemeAdwaita.cpp:
* Source/WebCore/platform/adwaita/ThemeAdwaita.h:
* Source/WebCore/platform/mac/ThemeMac.h:
* Source/WebCore/platform/mac/ThemeMac.mm:
* Source/WebCore/rendering/RenderTheme.cpp:
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebCore/rendering/adwaita/RenderThemeAdwaita.cpp:
* Source/WebCore/rendering/adwaita/RenderThemeAdwaita.h:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
* Source/WebCore/rendering/mac/RenderThemeMac.h:
* Source/WebCore/rendering/mac/RenderThemeMac.mm:
    - Move controlSize() and minimumControlSize() to RenderTheme.

* Source/WebCore/dom/Document.cpp:
    - Removes variable `hasSpecifiedLogicalHeight` which was always false due to the expression
      `renderer->style().logicalMaxHeight().isAuto()` always being false. This was caught statically
      by the fact that `logicalMaxHeight()` now returns a `Style::MaximumSize` which has no `isAuto()`
      function on it.

* Source/WebCore/html/shadow/ProgressShadowElement.cpp:
    - Don't allow negative size values to be set here as the property is defined to take nonnegative lengths only.

* Source/WebCore/html/NumberInputType.cpp:
* Source/WebCore/html/SearchInputType.cpp:
* Source/WebCore/html/HTMLInputElement.cpp:
* Source/WebCore/html/HTMLTextFormControlElement.cpp:
* Source/WebCore/html/shadow/DateTimeNumericFieldElement.cpp:
* Source/WebCore/html/shadow/DateTimeSymbolicFieldElement.cpp:
* Source/WebCore/html/shadow/TextControlInnerElements.cpp:
* Source/WebCore/layout/formattingContexts/FormattingGeometry.cpp:
* Source/WebCore/layout/formattingContexts/FormattingGeometry.h:
* Source/WebCore/layout/formattingContexts/block/BlockFormattingGeometry.cpp:
* Source/WebCore/layout/formattingContexts/block/BlockFormattingQuirks.cpp:
* Source/WebCore/layout/formattingContexts/block/BlockMarginCollapse.cpp:
* Source/WebCore/layout/formattingContexts/flex/FlexFormattingContext.cpp:
* Source/WebCore/layout/formattingContexts/table/TableFormattingContext.cpp:
* Source/WebCore/layout/formattingContexts/table/TableFormattingQuirks.cpp:
* Source/WebCore/layout/integration/flex/LayoutIntegrationFlexLayout.cpp:
* Source/WebCore/layout/layouttree/LayoutElementBox.cpp:
* Source/WebCore/layout/layouttree/LayoutTreeBuilder.cpp:
* Source/WebCore/page/LocalFrameView.cpp:
* Source/WebCore/rendering/AutoTableLayout.cpp:
* Source/WebCore/rendering/AutoTableLayout.h:
* Source/WebCore/rendering/FixedTableLayout.cpp:
* Source/WebCore/rendering/FixedTableLayout.h:
* Source/WebCore/rendering/GridLayoutFunctions.cpp:
* Source/WebCore/rendering/GridMasonryLayout.cpp:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:
* Source/WebCore/rendering/PositionedLayoutConstraints.cpp:
* Source/WebCore/rendering/RenderAttachment.cpp:
* Source/WebCore/rendering/RenderBlock.cpp:
* Source/WebCore/rendering/RenderBlockFlow.cpp:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
* Source/WebCore/rendering/RenderBoxModelObject.h:
* Source/WebCore/rendering/RenderButton.cpp:
* Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp:
* Source/WebCore/rendering/RenderElement.cpp:
* Source/WebCore/rendering/RenderFileUploadControl.cpp:
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
* Source/WebCore/rendering/RenderFlexibleBox.h:
* Source/WebCore/rendering/RenderFragmentContainer.cpp:
* Source/WebCore/rendering/RenderGrid.cpp:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderListBox.cpp:
* Source/WebCore/rendering/RenderListMarker.cpp:
* Source/WebCore/rendering/RenderMenuList.cpp:
* Source/WebCore/rendering/RenderMultiColumnSet.cpp:
* Source/WebCore/rendering/RenderObject.cpp:
* Source/WebCore/rendering/RenderReplaced.cpp:
* Source/WebCore/rendering/RenderScrollbarPart.cpp:
* Source/WebCore/rendering/RenderSlider.cpp:
* Source/WebCore/rendering/RenderTable.cpp:
* Source/WebCore/rendering/RenderTable.h:
* Source/WebCore/rendering/RenderTableCell.cpp:
* Source/WebCore/rendering/RenderTableCell.h:
* Source/WebCore/rendering/RenderTableCellInlines.h:
* Source/WebCore/rendering/RenderTableSection.cpp:
* Source/WebCore/rendering/RenderTableSection.h:
* Source/WebCore/rendering/RenderTextControl.cpp:
* Source/WebCore/rendering/RenderTextControlMultiLine.cpp:
* Source/WebCore/rendering/RenderTextControlSingleLine.cpp:
* Source/WebCore/rendering/line/BreakingContext.h:
* Source/WebCore/rendering/mathml/RenderMathMLBlock.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleBoxData.cpp:
* Source/WebCore/rendering/style/StyleBoxData.h:
* Source/WebCore/rendering/style/StyleFlexibleBoxData.cpp:
* Source/WebCore/rendering/style/StyleFlexibleBoxData.h:
* Source/WebCore/rendering/style/TextSizeAdjustment.cpp:
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.cpp:
* Source/WebCore/rendering/updating/RenderTreeBuilder.cpp:
* Source/WebCore/rendering/updating/RenderTreeUpdaterViewTransition.cpp:
* Source/WebCore/style/StyleAdjuster.cpp:
* Source/WebCore/svg/SVGLengthContext.cpp:
* Source/WebCore/svg/SVGLengthContext.h:
    - Update for new types.

Canonical link: https://commits.webkit.org/296045@main
